### PR TITLE
Add CPU Metrics using Node-Exporter and setup Pushgateway for Prometheus

### PR DESCRIPTION
**Pull request by burnerlee**
Add node exporter container and add it as a scrape target for prometheus. We are able to query for various CPU metrics that we need using prometheus and grafana now. Ready for deployment on staging. 
Setup a push gateway mechanism for prometheus in django and add a sample push metrics. Currently the sample push gateway mechanism is added for `get_all_challenges` api. Thus when a user hits this url, then the counter for http requests for this api is increased and the metric is pushed to pushgateway which can then be scraped by prometheus.
The counter is defined as total_http_requests which can be queried for the 4 arguments in any way: `django_app_name`,`endpoint_name`,`request_type`,`response_code`
memoryusage
pushgateway

### CODE REVIEW

**Comment by burnerlee** [comment]
Currently working on fixing the failing tests for the added custom metrics push in view.py for challenge

File: apps/challenges/views.py [hunk 1]
```diff
@@ -423,6 +424,10 @@ def get_all_challenges(request, challenge_time):
     # make sure that a valid url is requested.
     if challenge_time.lower() not in ("all", "future", "past", "present"):
         response_data = {"error": "Wrong url pattern!"}
+        Metrics.total_http_requests_sample.labels(
+            "challenge", "get_all_challenges", request.method, "406"
```
Comment (Ram81) [inline apps/challenges/views.py:428]: @burnerlee we need to move this to a utils method which will accept `app_label` `api_label` `request_type` and `status`

File: apps/challenges/views.py [hunk 2]
```diff
@@ -446,6 +451,10 @@ def get_all_challenges(request, challenge_time):
         result_page, many=True, context={"request": request}
     )
     response_data = serializer.data
+    Metrics.total_http_requests_sample.labels(
+        "challenge", "get_all_challenges", request.method, "200"
+    ).inc()
```
Comment (Ram81) [inline apps/challenges/views.py:456]: Also, we need to use 1 counter for each API and use tags in prometheus to distinguish them. This allows easy grouping and queries on a metric

File: apps/metrics/common.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
+from prometheus_client import Counter, REGISTRY, push_to_gateway
+
+
+class Metrics:
+    total_http_requests_sample = Counter(
+        "total_http_requests_sample",
+        "Total number of http requests completed",
```
Comment (Ram81) [inline apps/metrics/common.py:7]: This should be `Total number of http requests`

File: docker/prod/nodejs/nginx_staging.conf [hunk 4]
```diff
@@ -2,6 +2,14 @@ upstream django_app {
   server django:8000 fail_timeout=0;
 }
 
+upstream node_exporter {
+  server nodeexporter:9100 fail_timeout=0;
+}
+
+upstream prometheus_pushgateway {
+  server pushgateway:9091 fail_timeout=0;
```
Comment (Ram81) [inline docker/prod/nodejs/nginx_staging.conf:10]: We need only the node exporter on the same instance as EvalAI right?

**Review by Ram81 — CHANGES_REQUESTED** [review]

File: docker/prod/nodejs/nginx_staging.conf (hunk 4 above)
Comment (burnerlee) [reply docker/prod/nodejs/nginx_staging.conf:10]: @Ram81 yes

File: apps/jobs/metrics.py [hunk 5]
```diff
@@ -0,0 +1,6 @@
+from prometheus_client import Counter
+
+submissions_loaded_in_queue = Counter(
```
Comment (Ram81) [inline apps/jobs/metrics.py:3]: @burnerlee this is not correct. We need a wrapper on `Counter` class that will have an `inc` method. This wrapper will create a prometheus counter object with tags and labels, etc. All we have to do in an API is create one object of the Counter wrapper and use it to make `increment` calls for tags

File: apps/jobs/sender.py [hunk 6]
```diff
@@ -83,6 +84,8 @@ def publish_submission_message(message):
     queue_name = challenge.queue
     slack_url = challenge.slack_webhook_url
     queue = get_or_create_sqs_queue(queue_name)
+    # increase metrics counter for messages pushed in a queue
+    submissions_loaded_in_queue.inc()
```
Comment (Ram81) [inline apps/jobs/sender.py:88]: please make the required changes

File: docker/prod/docker_staging.env [hunk 7]
```diff
@@ -55,3 +55,10 @@ EKS_NODEGROUP_ROLE_ARN=x
 SLACK_WEB_HOOK_URL=x
 ENVIRONMENT=staging
 SERVICE_DNS=x
+
+GF_SECURITY_ADMIN_USER=x
+GF_SECURITY_ADMIN_PASSWORD=x
+
+prometheus_multiproc_dir=/code/monitoring/prometheus/django
+
+PUSHGATEWAY_ENDPOINT=https://staging.eval.ai/pushgateway
```
Comment (Ram81) [inline docker/prod/docker_staging.env:64]: Add newline. Don't hardcode the endpoint here. We don't check in endpoints in `_staging` or `_prod` `.env` files

File: scripts/workers/submission_worker.py [hunk 8]
```diff
@@ -676,9 +676,29 @@ def process_submission_callback(body):
                 SUBMISSION_LOGS_PREFIX, body
             )
         )
+        submission_pk = json.loads(body)["submission_pk"]
         body = yaml.safe_load(body)
         body = dict((k, int(v)) for k, v in body.items())
-        process_submission_message(body)
+        pushgateway_endpoint = os.environ.get("PUSHGATEWAY_ENDPOINT")
+        response = requests.post(
+            "{endpoint}/metrics/job/{j}/instance/{i}/submission_pk/{s}".format(
```
Comment (Ram81) [inline scripts/workers/submission_worker.py:684]: It is not a good practice to just put the request in the main block. Please create a separate method for this counter update. And why are we making a explicit post request for deque counting? We should use Prometheus client.

File: scripts/workers/submission_worker.py [hunk 9]
```diff
@@ -676,9 +676,29 @@ def process_submission_callback(body):
...
+                endpoint=pushgateway_endpoint,
+                j="submission-worker",
+                i="pushgateway",
+                s=submission_pk,
+            ),
+            data="{k} {v}\n".format(k="submissions_unloaded_from_queue", v=1),
```
Comment (Ram81) [inline scripts/workers/submission_worker.py:690]: We can't count the submission as processed as soon as it is read from the queue. This should be done only after submission is processed successfully or marked failed

**Review by Ram81 — CHANGES_REQUESTED** [review]

File: scripts/workers/submission_worker.py (hunk 9 above)
Comment (burnerlee) [reply scripts/workers/submission_worker.py:690]: @Ram81 i thought we are monitoring the submissions currently in queue.. so shouldn't be decrease the counter just after message is unloaded from the queue?

File: scripts/workers/submission_worker.py (hunk 8 above)
Comment (burnerlee) [reply scripts/workers/submission_worker.py:684]: Prometheus Pushgateway does not provide a support for using multiprocess registries which have been used in our project due to multiple workers in django... so if I use prometheus client,I would need to specify a registry. If I don't use the multiprocess registry, the results would be inconsistent as each worker would have an instance of registry.. So after some debugging and reading, I switched to API calls which give consistent and accurate results

File: docker/prod/docker_staging.env (hunk 7 above)
Comment (burnerlee) [reply docker/prod/docker_staging.env:64]: @Ram81 how should I store the url for pushgateway? i would need to access it in the submission_worker.py

File: docker/dev/docker.env [hunk 10]
```diff
@@ -35,3 +35,7 @@ SERVICE_DNS=localhost
 
 GF_SECURITY_ADMIN_USER=admin    
 GF_SECURITY_ADMIN_PASSWORD=password
+
+prometheus_multiproc_dir=/code/monitoring/prometheus/django
+
+PUSHGATEWAY_ENDPOINT=http://192.168.29.33/pushgateway
```
Comment (Ram81) [inline docker/dev/docker.env:41]: Why is the pushgateway IP hardcoded here?

File: docker/prod/docker_staging.env (hunk 7 above)
Comment (Ram81) [reply docker/prod/docker_staging.env:64]: Just mark it x for now. Env files are pulled from S3 where we store original values

File: scripts/workers/submission_worker.py (hunk 9 above)
Comment (Ram81) [reply scripts/workers/submission_worker.py:690]: No, we are monitoring number of unprocessed submissions. If a submission is not marked success/failed it is not considered as processed

File: scripts/workers/submission_worker.py (hunk 8 above)
Comment (Ram81) [reply scripts/workers/submission_worker.py:684]: @burnerlee for workers you don't have to worry about multiprocessing. We only have 1 worker per challenge at the moment. We can add support for multiprocessing later

File: scripts/workers/submission_worker.py [hunk 11]
```diff
@@ -676,9 +676,29 @@ def process_submission_callback(body):
...
+        )
+        if response.status_code != 200:
```
Comment (Ram81) [inline scripts/workers/submission_worker.py:692]: Dont use hardcoded status values. Use `HttpStatus.OK` instead

File: scripts/workers/submission_worker.py [hunk 12]
```diff
@@ -676,9 +676,29 @@ def process_submission_callback(body):
...
+            logger.exception(
+                "{} Exception while pushing metrics to pushgateway with status code {} due to error {}".format(
```
Comment (Ram81) [inline scripts/workers/submission_worker.py:694]: Change to `{} Exception when pushing metrics to push gateway: {}`

**Review by Ram81 — CHANGES_REQUESTED** [review]

File: scripts/workers/submission_worker.py [hunk 13]
```diff
@@ -669,16 +670,39 @@ def process_add_challenge_message(message):
     extract_challenge_data(challenge, phases)
 
 
-def process_submission_callback(body):
+def push_metrics_to_pushgateway(submission_pk, queue_name):
+    try:
+        registry = CollectorRegistry()
+        submissions_unloaded_from_queue = Counter(
```
Comment (Ram81) [inline scripts/workers/submission_worker.py:676]: Why is this a local counter? This should be global counter right?

File: apps/jobs/metrics.py [hunk 14]
```diff
@@ -0,0 +1,18 @@
+import os
+
+from prometheus_client import Counter, pushadd_to_gateway, CollectorRegistry
+
+
+pushgateway_registry = CollectorRegistry()
+
+submissions_loaded_in_queue = Counter(
```
Comment (Ram81) [inline apps/jobs/metrics.py:8]: Rename the variable to `num_submission_in_queue`

File: scripts/workers/submission_worker.py (hunk 13 above)
Comment (Ram81) [inline scripts/workers/submission_worker.py:676]: Rename the variable to `num_processed_submissions`

File: scripts/workers/submission_worker.py [hunk 15]
```diff
@@ -669,16 +670,39 @@ def process_add_challenge_message(message):
...
         logger.info(
             "{} [x] Received submission message {}".format(
                 SUBMISSION_LOGS_PREFIX, body
             )
         )
+        submission_pk = json.loads(body)["submission_pk"]
         body = yaml.safe_load(body)
         body = dict((k, int(v)) for k, v in body.items())
         process_submission_message(body)
+        push_metrics_to_pushgateway(submission_pk, queue_name)
```
Comment (Ram81) [inline scripts/workers/submission_worker.py:705]: Why are we adding `push_metric` call here? This should be done after message is deleted from the queue. Please make the required changes

File: scripts/workers/submission_worker.py [hunk 16]
```diff
@@ -831,7 +855,7 @@ def main():
                             WORKER_LOGS_PREFIX, message.body
                         )
                     )
-                    process_submission_callback(message.body)
+                    process_submission_callback(message.body, queue_name)
```
Comment (Ram81) [inline scripts/workers/submission_worker.py:858]: No need to pass `queue_name` to the submission processing callback

File: scripts/workers/submission_worker.py [hunk 17]
```diff
@@ -813,7 +837,7 @@ def main():
...
```
Comment (Ram81) [inline scripts/workers/submission_worker.py:840]: No need to pass `queue_name` to the submission processing callback

**Review by Ram81 — CHANGES_REQUESTED** [review]

File: docker-compose-staging.yml [hunk 18]
```diff
@@ -130,3 +129,17 @@ services:
     ports:
       - '9125:9125'
       - '9102:9102'
+  
+  node-exporter:
+    hostname: nodeexporter
```
Comment (Ram81) [inline docker-compose-staging.yml:134]: Can you rename the hostname to `node_exporter`

File: docker-compose-staging.yml [hunk 19]
```diff
@@ -130,3 +129,17 @@ services:
...
+    image: prom/node-exporter
+    ports:
+      - '9100:9100'
+
+  pushgateway:
+    hostname: pushgateway
```
Comment (Ram81) [inline docker-compose-staging.yml:140]: same for this hostname and endpoint

File: docker-compose.yml [hunk 20]
```diff
@@ -101,3 +101,17 @@ services:
     ports:
       - '9125:9125'
       - '9102:9102'
+      
+  node-exporter:
+    hostname: nodeexporter
```
Comment (Ram81) [inline docker-compose.yml:106]: same changes here

File: docker-compose.yml [hunk 21]
```diff
@@ -101,3 +101,17 @@ services:
...
+    image: prom/node-exporter
+    ports:
+      - '9100:9100'
+
+  pushgateway:
+    hostname: pushgateway
```
Comment (Ram81) [inline docker-compose.yml:112]: same changes here

File: docker/dev/docker.env [hunk 22]
```diff
@@ -38,3 +38,4 @@ GF_SECURITY_ADMIN_PASSWORD=password
 
 STATSD_ENDPOINT=statsd
 STATSD_PORT=9125
+PUSHGATEWAY_ENDPOINT=http://pushgateway:9091/pushgateway
```
Comment (Ram81) [inline docker/dev/docker.env:41]: update the endpoint here as well

File: docker/prod/nodejs/nginx_staging.conf [hunk 23]
```diff
@@ -6,6 +6,14 @@ upstream statsd_exporter {
   server statsd:9102 fail_timeout=0;
 }
 
+upstream node_exporter {
+  server nodeexporter:9100 fail_timeout=0;
+}
+
+upstream prometheus_pushgateway {
```
Comment (Ram81) [inline docker/prod/nodejs/nginx_staging.conf:13]: just use `push_gateway` here

File: docker/prod/nodejs/nginx_staging.conf [hunk 24]
```diff
@@ -64,6 +72,20 @@ server {
...
   }
 
+  location /nodeexporter {
+    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
+    proxy_set_header Host $host;
+    proxy_set_header X-Forwarded-Proto $scheme;
+    proxy_pass http://node_exporter/metrics;
+  }
+
+  location /pushgateway {
```
Comment (Ram81) [inline docker/prod/nodejs/nginx_staging.conf:82]: change route here as well

File: docker/prod/nodejs/nginx_staging.conf [hunk 25]
```diff
@@ -64,6 +72,20 @@ server {
     proxy_pass http://statsd_exporter;
   }
 
+  location /nodeexporter {
```
Comment (Ram81) [inline docker/prod/nodejs/nginx_staging.conf:75]: change route here as well

File: docker/prod/nodejs/nginx_staging.conf [hunk 26]
```diff
@@ -64,6 +72,20 @@ server {
...
+    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
+    proxy_set_header Host $host;
+    proxy_set_header X-Forwarded-Proto $scheme;
+    proxy_pass http://node_exporter/metrics;
```
Comment (Ram81) [inline docker/prod/nodejs/nginx_staging.conf:79]: be consistent with your naming convention. At some place you are using `node_exporter` and at other places `nodeexporter`. This is not a good practice

File: docker/prod/nodejs/nginx_staging.conf [hunk 27]
```diff
@@ -64,6 +72,20 @@ server {
...
+  }
+
+  location /pushgateway {
+    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
+    proxy_set_header Host $host;
+    proxy_set_header X-Forwarded-Proto $scheme;
+    proxy_pass http://prometheus_pushgateway;
```
Comment (Ram81) [inline docker/prod/nodejs/nginx_staging.conf:86]: rename upstream

File: docker/prod/nodejs_v2/nginx_staging.conf [hunk 28]
```diff
@@ -44,4 +52,18 @@ server {
...
+    proxy_set_header X-Forwarded-Proto $scheme;
+    proxy_pass http://node_exporter/metrics;
+  }
+
+  location /pushgateway {
+    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
+    proxy_set_header Host $host;
+    proxy_set_header X-Forwarded-Proto $scheme;
+    proxy_pass http://prometheus_pushgateway;
+  }
```
Comment (Ram81) [inline docker/prod/nodejs_v2/nginx_staging.conf:68]: make same changes here as well

**Review by Ram81 — CHANGES_REQUESTED** [review]

File: docker-compose-staging.yml [hunk 29]
```diff
@@ -130,3 +129,17 @@ services:
...
       - '9125:9125'
       - '9102:9102'
+  
+  node-exporter:
+    hostname: node_exporter
+    image: prom/node-exporter
+    ports:
+      - '9100:9100'
+
+  pushgateway:
```
Comment (Ram81) [inline docker-compose-staging.yml:139]: @burnerlee this is inconsitent

File: docker-compose-staging.yml [hunk 30]
```diff
@@ -130,3 +129,17 @@ services:
     ports:
       - '9125:9125'
       - '9102:9102'
+  
+  node-exporter:
```
Comment (Ram81) [inline docker-compose-staging.yml:133]: @burnerlee this is inconsitent. We use `_` as separator for components not `-`.

File: docker-compose.yml [hunk 31]
```diff
@@ -101,3 +101,17 @@ services:
     ports:
       - '9125:9125'
       - '9102:9102'
+      
+  node-exporter:
```
Comment (Ram81) [inline docker-compose.yml:105]: same here

File: docker-compose.yml [hunk 32]
```diff
@@ -101,3 +101,17 @@ services:
...
+    hostname: node_exporter
+    image: prom/node-exporter
+    ports:
+      - '9100:9100'
+
+  pushgateway:
```
Comment (Ram81) [inline docker-compose.yml:111]: same here

File: monitoring/prometheus/prometheus.yml [hunk 33]
```diff
@@ -11,3 +11,13 @@ scrape_configs:
     metrics_path: '/statsd/metrics'
     static_configs:
       - targets: ['statsd:9102']
+  
+  - job_name: 'node-exporter'
```
Comment (Ram81) [inline monitoring/prometheus/prometheus.yml:15]: Fix the name here too

File: monitoring/prometheus/prometheus.yml [hunk 34]
```diff
@@ -11,3 +11,13 @@ scrape_configs:
...
+    metrics_path: '/metrics'
+    static_configs:
+      - targets: ['node_exporter:9100']
+  
+  - job_name: 'pushgateway'
```
Comment (Ram81) [inline monitoring/prometheus/prometheus.yml:20]: Here too

File: monitoring/prometheus/prometheus_staging.yml [hunk 35]
```diff
@@ -11,4 +11,14 @@ scrape_configs:
   - job_name: 'statsd'
     metrics_path: '/statsd/metrics'
     static_configs:
-      - targets: ['https://staging.eval.ai']
+      - targets: ['staging.eval.ai']
+
+  - job_name: 'node-exporter'
```
Comment (Ram81) [inline monitoring/prometheus/prometheus_staging.yml:16]: Same here

File: monitoring/prometheus/prometheus_staging.yml [hunk 36]
```diff
@@ -11,4 +11,14 @@ scrape_configs:
...
+    metrics_path: '/node_exporter'
+    static_configs:
+      - targets: ['staging.eval.ai']
+  
+  - job_name: 'pushgateway'
```
Comment (Ram81) [inline monitoring/prometheus/prometheus_staging.yml:21]: same here

File: scripts/deployment/deploy.sh [hunk 37]
```diff
@@ -150,6 +150,16 @@ case $opt in
             docker-compose -f docker-compose-${env}.yml up -d statsd-exporter
             echo "Completed deploy operation."
             ;;
+        deploy-node-exporter)
+            echo "Deploying node-exporter docker container..."
+            docker-compose -f docker-compose-${env}.yml up -d node-exporter
+            echo "Completed deploy operation."
+            ;;
+        deploy-pushgateway)
```
Comment (Ram81) [inline scripts/deployment/deploy.sh:158]: Here you can use `-`

**Review by Ram81 — CHANGES_REQUESTED** [review]

File: scripts/deployment/deploy.sh [hunk 38]
```diff
@@ -150,6 +150,16 @@ case $opt in
             docker-compose -f docker-compose-${env}.yml up -d statsd-exporter
             echo "Completed deploy operation."
             ;;
+        deploy-node-exporter)
```
Comment (Ram81) [inline scripts/deployment/deploy.sh:153]: @burnerlee if push gateway and node exporter is part of nginx config it needs to be deployed with EvalAI backend and frontend. Please make the changes in `auto_deploy` command to start node exporter and push gateway. See line 50 for reference

**Review by Ram81 — CHANGES_REQUESTED** [review]

**Review by Ram81 — APPROVED** [review]
LGTM

**Review by RishabhJain2018 — APPROVED** [review]

**Pull request merged**